Roll over no longer active status into new academic year

rmainwork opened the pull request:
This PR builds on the work in #546 and cascades these changes into the new semester upon creation. This means that:

- A faculty member who is `ABSENCE_TYPE.NO_LONGER_ACTIVE` will cascade into the newly created semester as such
- A faculty member who is anything else will be set as `ABSENCE_TYPE.PRESENT` in the newly created semester.

## Types of changes
<!--
  What types of changes does your code introduce? Put an `x` in all the boxes
  that apply:
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality
      to change)

## Checklist:
<!--
  Go over all the following points, and put an `x` in all the boxes that apply.
-->
- [x] I have run `eslint` on the code
- [ ] I have added JSDoc for all of my code (where applicable)
- [x] I have added tests to cover my changes.

## Priority:
- [x] Normal <!-- New piece of functionality -->
- [ ] High <!-- Critical bug requiring urgent review -->

## Related Issues:
<!--
  Add the issue number in below that this PR is intended to address.

  Also mention any issues that this PR is related to, and if possible,
  provide more information regarding the relationship to the issues in
  question (i.e does the PR fix the issue, is it designed to fix another bug
  that is similar to the issue etc.)
-->
Fixes #534

<!--
  Attribution:
  PR Template adapted from: https://github.com/h5bp/html5-boilerplate/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

Commit:
Insert NO_LONGER_ACTIVE absences in new academic year

If the faculty member was previously `NO_LONGER_ACTIVE`, copy that into
the new academic year, otherwise set them to `PRESENT`.

Commit:
Remove spread operator

Usage of the spread operator here was causing the new absence to be
(incorrectly) linked back to the previous absence record which then
resulted in the absence record being updated (instead of a new one
created). This then cascaded into test failures as other values in the
database were different than expected as the code modified existing data
rather than creating new records.

jonseitz commented on tests/integration/server/semester/semester.service.test.ts:
```diff
@@ -377,6 +383,62 @@ describe('Semester Service', function () {
+            .leftJoinAndSelect('f.absences', 'absence')
+            .leftJoinAndSelect('absence.semester', 'semester')
+            .where({ id })
+            .andWhere(
+              'semester.academicYear = :acyr',
+              { acyr: testAcademicYear }
+            )
```
`semester.academicYear` is actually the calendar year while the `addAcademicYear` function is creating Spring in `year` and Fall in `year - 1`. 
So assuming that `testAcademicYear` is 2024, this would create Fall 2023 and Spring 2024, but the `WHERE` clause here would only return the "Spring 2024" semester.

I'd suggest duplicating the test format used above, and query `absenceService` to get the specific spring and fall absences associated with the faculty member. Then you can have separate assertions for each absences that check if `type === ABSENCE_TYPE.NO_LONGER_ACTIVE`

jonseitz commented on tests/integration/server/semester/semester.service.test.ts:
```diff
@@ -377,6 +383,62 @@ describe('Semester Service', function () {
+            .getOne();
+          await semesterService.addAcademicYear(testAcademicYear);
+
+          const { absences } = await facultyRepository
+            .createQueryBuilder('f')
+            .leftJoinAndSelect('f.absences', 'absence')
+            .leftJoinAndSelect('absence.semester', 'semester')
+            .where({ id: presentFacultyMember.id })
+            .andWhere('semester.academicYear = :acyr', { acyr: testAcademicYear })
+            .getOne();
+          const allPresent = absences
+            .every(({ type }) => type === ABSENCE_TYPE.PRESENT);
+          strictEqual(allPresent, true);
```
Notes on previous test block apply here as well.

jonseitz reviewed (changes requested):
There are a couple of issues with the tests, but the functionality here seems to be correct

jonseitz committed:
Test individual semesters from the new year

The previous tests were:

  1. Only checking the Spring semester, as the fall semester has a
     different academic year.

  2. Dangerously checking the result of Array.every(), which would
     return true with an empty array. It would also only show true !==
     false in the event of a failing test.

This new format affirmatively checks that the type of each absence is
NO_LONGER_ACTIVE or PRESENT, giving more direct feedback about what
would have gone wrong in a failure.

jonseitz committed:
Add a one-time fix for missing academic year

When we copied the data from the old Course Planner there were no
semesters for the 2025-2026 academic year. This would be the fourth year
of our four-year plan in 2023, which causes two problems:

1. The admins are unable to add anything for this year, leaving the
   columns blank in the interface.
2. There will no "previous year" data to copy over when we try to add
   the next new year (2026-2027) in June.

To resolve this, we can add a hard-coded call to addAcademicYear that
will add the missing 2026 year. Because this won't run if the year
already exists, this shouldn't cause problems if it attempts to call the
function twice. However, we should probably just revert this commit and
drop the extra code after the has been populated.

jonseitz committed:
Add new academic years any time after June

I think originally we were concerned about running the process too many
times and potentially slowing down the container startup, but I think
that's a less significant concern that potentially skipping a year if
the container doesn't restart in June. I'm more confident that we'll
restart the container at least once during the back half of the year,
and we can monitor logs from that period to make sure we're not causing
unintentional slowdowns or other issues.

natalynyu commented on tests/integration/server/semester/semester.service.test.ts:
```diff
@@ -377,6 +383,92 @@ describe('Semester Service', function () {
+      });
+      describe(`is anything other than ${ABSENCE_TYPE.NO_LONGER_ACTIVE}`, function () {
+        it(`defaults to ${ABSENCE_TYPE.PRESENT}`, async function () {
+          const presentFacultyMember = await facultyRepository
+            .createQueryBuilder('f')
+            .leftJoinAndSelect('f.absences', 'absence')
+            .where('absence.type = :type', { type: ABSENCE_TYPE.PARENTAL_LEAVE })
+            .getOne();
```
I think it may be a good idea to rename this variable to `activeFacultyMember` instead of `presentFacultyMember`, since this is supposed to be a faculty member on parental leave.

jonseitz replied:
Good point, that lines up better with the test case overall. I made that change in c73bfdd

natalynyu reviewed (changes requested):
This all seems to be in working order - I just had a small note about a test variable.

jonseitz committed:
Rename facultyMember variable

"Active" is a better description for this test case, and we're mainly
testing the behavior when the faculty is not "No Longer Active"

natalynyu reviewed (approved):
Approved!
